### ISSUE TITLE
"Diff lag" when data resource depends on -/+ resource

### ISSUE THREAD

**Issue by apparentlymart** [hashicorp/terraform#6824]
Consider the following configuration:

```hcl
resource "aws_vpc" "main" {
  # ...
}

# Useless to have this in the same module as the resource, but imagine it's in a separate module
data "aws_vpc" "main" {
  id = "${aws_vpc.main.id}"
}

resource "aws_subnet" "a" {
  vpc_id = "${data.aws_vpc.main.id}"
  # ...
}
```

On first create the above works as expected: the VPC is created first, then its data is retrieved from the AWS API by the data source, and then the subnet is created.

The problem arises when a change is made to the VPC resource that forces it to be replaced with a new resource. In that case:

* Refreshing `data.aws_vpc.main` is refreshed during the refresh walk, using the old VPC id that's still present in the state.
* A `-/+` diff is produced for `aws_vpc.main`, causing `aws_vpc.main.id` to be `<computed>`
* ...but we already refreshed the `data.aws_vpc.main` during the refresh walk, so we don't produce a further read diff for it at this point, causing the old VPC id to persist
* `aws_subnet.a` gets interpolated from the current state of `data.aws_vpc.main.id`, which is still the old VPC id

Consequently, it takes two plan/apply cycles to fully propagate the change through.

The expected behavior would be for us to notice on the plan walk that `${aws_vpc.main.id}` became computed in the mean time and thus we need to do place a read diff into the plan even though there is an instance already in the state.

I expect this requires some extra/different steps in the `EvalTree` implementation for data resources on `walkPlan`. Currently it just simplistically says "if we already have a state then we don't need to plan at all", but in fact we need to re-interpolate the configuration here and see if any computed values are present, and ignore the existing state and diff again if so.

**Comment by apparentlymart** [comment]
On reflection I'm considering that maybe the problem here is just an incorrect conception of the problem.

In the implementation that was merged, the rules were:
* During refresh, if a data resource configuration is completely known, immediately read it
* During plan, if a data resource is lacking a primary `InstanceState`, generate a diff for it
* During apply, if a diff is present for a data resource then read it

I think the correct flow is actually the following:
* During refresh, if a data resource configuration is completely known, immediately read it
* During plan, if a data resource configuration is *not* completely known, generate a diff for it
* During apply, if a diff is present for a data resource then read it

In other words, the generation of the diff should be contingent on whether the configuration is computed, not on whether any instance happens to already be present in the state. That way a configuration that is non-computed during refresh (because it's using managed resource attributes from a previous run) can become computed during plan (because a managed resource has been marked for replacement).
